core/modules.rs

```diff
@@ -0,0 +1,436 @@
+    &mut self,
+    specifier: &str,
+    referrer: &str,
+    parent_id: Option<deno_mod>,
+  ) {
+    let url = {
+      let loader = self.loader.as_mut().unwrap();
+      loader.resolve(specifier, referrer)
+    };
+
+    if let Some(parent_id) = parent_id {
+      let loader = self.loader.as_mut().unwrap();
+      loader.use_modules(|modules| modules.add_child(parent_id, &url));
+    }
+
+    if !self.is_pending.contains(&url) {
+      self.is_pending.insert(url.clone());
+      let source_code_future = {
+        let loader = self.loader.as_mut().unwrap();
+        Box::new(loader.load(&url))
+      };
+      self.pending.push(PendingLoad {
+        url,
+        source_code_future,
+      });
+    }
+  }
+}
+
+// TODO(ry) This is basically the same thing as RustOrJsError. They should be
+// combined into one type.
+pub enum Either<E> {
+  JSError(JSError),
+  Other(E),
+}
+
+// TODO remove 'static below.
+impl<E: 'static + Error, B: Behavior, L: 'static + Loader<E, B>> Future
+  for RecursiveLoad<E, B, L>
+{
+  type Item = (deno_mod, L);
+  type Error = (Either<E>, L);
+
+  fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+    let mut i = 0;
+    while i < self.pending.len() {
+      let pending = &mut self.pending[i];
+      match pending.source_code_future.poll() {
+        Err(err) => {
+          return Err((Either::Other(err), self.take_loader()));
+        }
+        Ok(Async::NotReady) => {
+          i += 1;
+        }
+        Ok(Async::Ready(source_code)) => {
+          // We have completed loaded one of the modules.
+          let completed = self.pending.remove(i);
+          let main = completed.url == self.root;
+
+          let result = {
+            let loader = self.loader.as_mut().unwrap();
+            loader.use_isolate(|isolate: &mut Isolate<B>| {
+              isolate.mod_new(main, &completed.url, &source_code)
+            })
+          };
+          if let Err(err) = result {
+            return Err((Either::JSError(err), self.take_loader()));
+          }
+          let mod_id = result.unwrap();
+          let referrer = &completed.url.clone();
+
+          {
+            let loader = self.loader.as_mut().unwrap();
+            loader.use_modules(|modules: &mut Modules| {
+              modules.register(mod_id, &completed.url)
+            });
+          }
+
+          // Now we must iterate over all imports of the module and load them.
+          let imports = {
+            let loader = self.loader.as_mut().unwrap();
+            loader.use_isolate(|isolate| isolate.mod_get_imports(mod_id))
+          };
+          for specifier in imports {
+            self.add(&specifier, referrer, Some(mod_id));
+          }
+        }
+      }
+    }
+
+    if self.pending.len() > 0 {
+      return Ok(Async::NotReady);
+    }
+
+    let mut loader = self.take_loader();
+
+    // TODO Fix this resolve callback weirdness.
+    let loader_ =
+      unsafe { std::mem::transmute::<&mut L, &'static mut L>(&mut loader) };
```

Comment:
There are two problems here.

First is that isolate::ResolveFn is defined as `FnMut(...) -> deno_mod + 'static`. The 'static part is implicit. This is easy to fix.

The other problem is that if you mutably borrow the `isolate` field from MockLoader, the loader is locked thereafter and it's other fields are inaccessible. So the following pattern can't possibly work: 
```
loader.use_isolate(|isolate| {
  isolate.mod_instantiate(some_id, |specifier, referrer| {
    let id = loader.resolve(specifier, referred);   // <-- This doesn't work because `loader.use_isolate()` has made `loader` inaccessible.
    // ... do something with id.
 });
});
```

This is more difficult to fix and speaks against the design with use_modules()/use_isolate().

Reply:
Partially fixed by merging `use_isolate()` and `use_modules()` into one function `fn enter<R, F: FnMut(&mut Isolate<Self::Behavior>, &mut Modules) -> R>(&mut self, cb: F) -> R;` and making `resolve()` a static method.

Whether this is sufficient when doing the 'real' Loader implementation remains to be seen; it could be that you'll add more fields that you'll want to access while also having access to the isolate/modules.

Reply:
I've already done the real loader. I'm just breaking this up for ease of landing.